vm/actor/src/builtin/miner/bitfield_queue.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::{actor_error, ActorError, ExitCode};
+
+use super::QuantSpec;
+use bitfield::BitField;
+use cid::Cid;
+use clock::ChainEpoch;
+use ipld_amt::{Amt, Error as AmtError};
+use ipld_blockstore::BlockStore;
+use std::collections::HashMap;
+use std::error::Error as StdError;
+
+/// Wrapper for working with an AMT[ChainEpoch]*Bitfield functioning as a queue, bucketed by epoch.
+/// Keys in the queue are quantized (upwards), modulo some offset, to reduce the cardinality of keys.
+pub struct BitFieldQueue<'db, BS> {
+    pub amt: Amt<'db, BitField, BS>,
+    quant: QuantSpec,
+}
+
+impl<'db, BS: BlockStore> BitFieldQueue<'db, BS> {
+    pub fn new(store: &'db BS, root: &Cid, quant: QuantSpec) -> Result<Self, AmtError> {
+        Ok(Self {
+            amt: Amt::load(root, store)?,
+            quant,
+        })
+    }
+
+    /// Adds values to the queue entry for an epoch.
+    pub fn add_to_queue(
+        &mut self,
+        raw_epoch: ChainEpoch,
+        values: &BitField,
+    ) -> Result<(), Box<dyn StdError>> {
+        if values.is_empty() {
+            // nothing to do.
+            return Ok(());
+        }
+
+        let epoch = self.quant.quantize_up(raw_epoch);
+
+        let bitfield = self
+            .amt
+            .get(epoch as u64)
+            .map_err(|e| format!("failed to lookup queue epoch {}: {:?}", epoch, e))?
+            .ok_or_else(|| actor_error!(ErrNotFound; ""))?;
+
+        self.amt
+            .set(epoch as u64, &bitfield | values)
+            .map_err(|e| format!("failed to set queue epoch {}: {:?}", epoch, e))?;
+
+        Ok(())
+    }
+
+    pub fn add_to_queue_values(
+        &mut self,
+        epoch: ChainEpoch,
+        values: &[u64],
+    ) -> Result<(), Box<dyn StdError>> {
+        self.add_to_queue(epoch, &values.iter().map(|&i| i as usize).collect())
+    }
+
+    /// Cut cuts the elements from the bits in the given bitfield out of the queue,
+    /// shifting other bits down and removing any newly empty entries.
+    ///
+    /// See the docs on `BitField::cut` to better understand what it does.
+    pub fn cut(&mut self, to_cut: &BitField) -> Result<(), String> {
+        let mut epochs_to_remove = Vec::<u64>::new();
+
+        self.amt
+            .for_each_mut(|epoch, bitfield| {
+                *bitfield = bitfield.cut(to_cut);
+
+                if bitfield.is_empty() {
+                    epochs_to_remove.push(epoch);
+                }
```

Comment:
There seems to me to be an issue which is that if the bitfield is empty, the indexed bitfield would still be updated when in specs-actors they only update if the resultant bitfield is not empty.

Is this the same logically and I'm just missing something about cut?

Reply:
I figured it doesn't matter because empty bitfields are removed from the AMT right after, so it's not important whether we make it empty before removing it or not. Does that make sense?